extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/output/RetriableS3OutputStream.java

```diff
@@ -0,0 +1,432 @@
+  private int numChunksPushed;
+  /**
+   * Total size of all chunks. This size is updated whenever the chunk is ready for push,
+   * not when {@link #write(byte[], int, int)} is called. This is because
+   * it will be hard to know the increase of chunk size in write() when the chunk is compressed.
+   */
+  private long resultsSize;
+
+  /**
+   * A flag indicating whether there was an upload error.
+   * This flag is tested in {@link #close()} to determine whether it needs to upload the current chunk or not.
+   */
+  private boolean error;
+  private boolean closed;
+
+  public RetriableS3OutputStream(
+      S3OutputConfig config,
+      ServerSideEncryptingAmazonS3 s3,
+      String s3Key
+  ) throws IOException
+  {
+
+    this(config, s3, s3Key, true);
+  }
+
+  @VisibleForTesting
+  protected RetriableS3OutputStream(
+      S3OutputConfig config,
+      ServerSideEncryptingAmazonS3 s3,
+      String s3Key,
+      boolean chunkValidation
+  ) throws IOException
+  {
+    this.config = config;
+    this.s3 = s3;
+    this.s3Key = s3Key;
+
+    final InitiateMultipartUploadResult result = s3.initiateMultipartUpload(
+        new InitiateMultipartUploadRequest(config.getBucket(), s3Key)
+    );
+    this.uploadId = result.getUploadId();
+    this.chunkStorePath = new File(config.getTempDir(), uploadId + UUID.randomUUID());
+    FileUtils.mkdirp(this.chunkStorePath);
+    this.chunkSize = config.getChunkSize() == null ? computeChunkSize(config) : config.getChunkSize();
+    if (chunkValidation) {
+      validateChunkSize(config.getMaxResultsSize(), chunkSize);
+    }
+    this.pushStopwatch = Stopwatch.createUnstarted();
+    this.pushStopwatch.reset();
+
+    this.currentChunk = new Chunk(nextChunkId, new File(chunkStorePath, String.valueOf(nextChunkId++)));
+  }
+
+  private static long computeChunkSize(S3OutputConfig config)
+  {
+    return computeMinChunkSize(config.getMaxResultsSize());
+  }
+
+  private static void validateChunkSize(long maxResultsSize, long chunkSize)
+  {
+    if (computeMinChunkSize(maxResultsSize) > chunkSize) {
+      throw new IAE(
+          "chunkSize[%s] is too small for maxResultsSize[%s]. chunkSize should be at least [%s]",
+          chunkSize,
+          maxResultsSize,
+          computeMinChunkSize(maxResultsSize)
+      );
+    }
+    if (S3_MULTIPART_UPLOAD_MAX_PART_SIZE < chunkSize) {
+      throw new IAE(
+          "chunkSize[%s] should be smaller than [%s]",
+          chunkSize,
+          S3_MULTIPART_UPLOAD_MAX_PART_SIZE
+      );
+    }
+  }
+
+  private static long computeMinChunkSize(long maxResultsSize)
+  {
+    return Math.max(
+        (long) Math.ceil(maxResultsSize / (double) S3_MULTIPART_UPLOAD_MAX_NUM_PARTS),
+        S3_MULTIPART_UPLOAD_MIN_PART_SIZE
+    );
+  }
+
+  public static String getS3KeyForQuery(String prefix, String asyncResultId)
+  {
+    return JOINER.join(
+        prefix,
+        asyncResultId
```

Comment:
Ah! So this is only for Async?

Since we're at a lower level, should this be generalized? For anything that wants to write to S3? And automate the math?